azurerm/resource_arm_cosmos_db_account.go

```diff
@@ -146,78 +226,213 @@ func resourceArmCosmosDBAccount() *schema.Resource {
+		}
+	} else {
+		//could be a CustomizeDiff?, but this is temporary
+		return fmt.Errorf("Neither `geo_location` or `failover_policy` is set for CosmosDB Account '%s'", name)
+	}
+
+	//get existing locations (if exists)
+	resp, err := client.Get(ctx, resourceGroup, name)
+	if err != nil {
+		return fmt.Errorf("Error making Read request on AzureRM CosmosDB Account '%s': %s", name, err)
+	}
+
+	oldLocations := []documentdb.Location{}
+	oldLocationsMap := map[string]documentdb.Location{}
+	for _, l := range *resp.FailoverPolicies {
+		location := documentdb.Location{
+			ID:               l.ID,
+			LocationName:     l.LocationName,
+			FailoverPriority: l.FailoverPriority,
+		}
+
+		oldLocations = append(oldLocations, location)
+		oldLocationsMap[azureRMNormalizeLocation(*location.LocationName)] = location
+	}
+
+	//cannot update properties and add/remove replication locations at the same time
+	//so first just update any changed properties
+	account := documentdb.DatabaseAccountCreateUpdateParameters{
 		Location: utils.String(location),
 		Kind:     documentdb.DatabaseAccountKind(kind),
 		DatabaseAccountCreateUpdateProperties: &documentdb.DatabaseAccountCreateUpdateProperties{
-			ConsistencyPolicy:        &consistencyPolicy,
-			Locations:                &failoverPolicies,
+			ConsistencyPolicy:        expandAzureRmCosmosDBAccountConsistencyPolicy(d),
+			Locations:                &oldLocations,
 			DatabaseAccountOfferType: utils.String(offerType),
 			IPRangeFilter:            utils.String(ipRangeFilter),
+			EnableAutomaticFailover:  utils.Bool(enableAutomaticFailover),
 		},
 		Tags: expandTags(tags),
 	}
 
-	future, err := client.CreateOrUpdate(ctx, resGroup, name, parameters)
-	if err != nil {
-		return err
+	if _, err := resourceArmCosmosDBAccountApiUpsert(client, ctx, resourceGroup, name, account); err != nil {
+		return fmt.Errorf("Error updating CosmosDB Account %q properties (Resource Group %q): %+v", name, resourceGroup, err)
 	}
 
-	err = future.WaitForCompletion(ctx, client.Client)
-	if err != nil {
-		return err
+	//determine if any locations have been renamed/priority reordered and remove them
+	removedOne := false
+	for _, l := range newLocations {
+		if ol, ok := oldLocationsMap[*l.LocationName]; ok {
+			if *l.FailoverPriority != *ol.FailoverPriority {
+				if *l.FailoverPriority == 0 {
+					return fmt.Errorf("Cannot change the failover priority of primary Cosmos DB account %q location %s to %d (Resource Group %q)", name, *l.LocationName, *l.FailoverPriority, resourceGroup)
+				}
+				delete(oldLocationsMap, *l.LocationName)
+				removedOne = true
+				continue
+			}
+			if *l.ID == "" && *ol.ID == resourceArmCosmosDBAccountGenerateDefaultId(name, *l.LocationName) {
+				continue
+			}
+			if *l.ID != *ol.ID {
+				if *l.FailoverPriority == 0 {
+					return fmt.Errorf("Cannot change the prefix/ID of the primary Cosmos DB account %q location %s (Resource Group %q)", name, *l.LocationName, resourceGroup)
+				}
```

Comment:
I'm assuming this is matching an Azure constraint?

Reply:
Yep, you can't change the `failover_priority` or `id` of an existing location.